### ISSUE TITLE
[question] [support needed] Ignore conent in smarty tags

### ISSUE THREAD

**Issue by hellboy81** [htmllint/htmllint#155]
I have HTML file with custom smarty tags included: `[{ .. }]`

How can I disable checking content/code inside Smarty tags?

For example

```
[{ $myObj->myMeth(...)}]
```

```
Special characters must be escaped : [ > ] // False positive
```

I am using `Atom` with `htmllint` and `language-smarty` plugins installed

**Comment by mlochbaum** [comment]
The easy way is to remove the tags before linting...

The main issue with these kinds of tags is that the parser (htmlparser2) might behave unpredictably, since they are not valid html. The tag you gave will always be output as a text element, but if it contained an opening bracket `<` instead of the closing one then it could output spurious tags or errors. The parser doesn't have fine enough control to prevent this from happening.

Because of those issues, the best completely reliable solution I can offer would be a `raw-ignore-regex` option which just removes everything matching a regex before parsing. If you use Smarty tags for things like attribute names or class values, this could cause errors resulting from those values being empty.

If the tags are always parsed as text, I can add an option to disable `text-escape-spec-char` on a specified regex. This would work just like existing `id-class-ignore-regex` and others. To my astonishment, `text-escape-spec-char` is the only rule that works on raw text, so I don't even have to worry about code duplication issues for the time being.

If you see any errors with different codes than one you gave (code E23), then option 2 is off the table. If you use Smarty inside opening and closing tags, then option 1 is not completely off the table but does require a lot of discussion. Not a great situation, but hopefully we can work something out.

**Comment by dharapvj** [comment]
Even I want to use the htmllint for underscore templates.. and i want to ignore anything comes inside `<% ... %>` tags.. `raw-ignore-regex` would certainly be handy..

**Comment by Potherca** [comment]
Any news on this?

**Comment by mlochbaum** [comment]
That last commit adds `text-ignore-regex` which is roughly option 2 above. It disables `spec-char-escape` within text matching the given regex. This fixes the issue described above as well as in #198. I would appreciate it if anyone having these problems could test it out.

I want to add a `raw-ignore-regex` as well, which just cuts out text before parsing. There are some issues to resolve here, however:
- Removing the text entirely breaks line/column numbers. Replacing it with a placeholder may cause some rules to behave improperly, e.g. `line-max-len`.
- Getting it to respect inline configuration is difficult. It's supposed to apply before parsing, but parsing is what tells us which bits of the file are inline configurations.
- It breaks the configuration structure where options belong to rules. Not insurmountable, just kind of ugly.

Version 1.0.0 goes out once this is fixed.

**Comment by mlochbaum** [comment]
Okay, finally found a solution I'm happy with for `raw-ignore-regex`. The option just strips out everything matching the given regex (it may be a good idea to allow replacing to avoid empty attributes, etc, but I'm keeping it simple for now). There's no inline configuration handling, since doing this without a quadratic slowdown requires halting parsing once we find a configuration tag; that would be hard to do. But careful use of the regex gives something similar: see [the relevant section in Options](https://github.com/htmllint/htmllint/wiki/Options#raw-ignore-regex).

This approach makes it clear that htmllint is not able to do anything magic with inline configuration, and it lets users "turn off" htmllint for a section.

It does break line and column numbers; I need to find a way to recover them.

**Comment by mlochbaum** [comment]
That was... harder than expected. Version 0.7.0 now out.
